compiler/rustc_lint/src/dangling.rs

```diff
@@ -151,6 +153,40 @@ fn lint_expr(cx: &LateContext<'_>, expr: &Expr<'_>) {
     }
 }
 
+/// Determine if a `#[rustc_as_ptr]` receiver type (with adjustments) is only applied
+/// to owned allocations.
+fn is_inner_allocation_owned(input_ty: Ty<'_>, adjs: &[Adjustment<'_>]) -> bool {
+    // 1. Look at all the deref-ed types and exclude any pointer types
+    //
+    //  - `Box<u8>` -> `u8`: OK
+    //  - `Box<&u8>` -> `&u8`: KO
+    //  - `Box<Box<u8>>` -> `Box<u8>` -> `u8`: OK
+    //  - `MaybeUninit<MaybeUninit<&u8>>`: OK
+    //  - `Box<MaybeUninit<&u8>>` -> `MaybeUninit<&u8>`: OK
+    adjs.iter()
+        .filter_map(|adj| match adj.kind {
+            Adjust::Deref(_) => Some(adj.target),
+            Adjust::NeverToAny
+            | Adjust::Borrow(_)
+            | Adjust::Pointer(_)
+            | Adjust::ReborrowPin(_) => None,
+        })
+        .all(|deref_ty| !deref_ty.is_any_ptr())
+        // 2. Look at the top level generics at exclude any pointer types
```

Comment:
I'm not sure we need this check at all. It seems to reject only 1 additional case, which does not need to be rejected in the first place.

Reply:
I have removed the first match arm, but the `!input_ty.is_any_ptr()` part is still neccessary to exclude types like starting with a reference or pointer.